### PR TITLE
add PubliBike (CH)

### LINKED ISSUE #337: Add PubliBike, Switzerland
There's a bike system across Switzerland that could be added to your listing : PubliBike https://www.publibike.ch/en/publibike
This is a complementary system, not replacing any previous system that I am aware of.

### PR DESCRIPTION
**Pull request by eUgEntOptIc44**
fix #337

Hi @eskerda 

As [mentioned](https://github.com/eskerda/pybikes/pull/514#issuecomment-1358657123) in #514 I looked into the possible integration of [*PubliBike*](https://www.publibike.ch/en/home) a bike sharing provider in Switzerland.

Unfortunately *PubliBike* does **not** offer to request only stations of a single *network*[^1]. Please also see their [OpenAPI](https://api.publibike.ch/v1/static/api.html) docs for more details.

Therefore I rely on the **caching** of `pybikes` and filter the api answer by the network's `id`. This allows to easily migrate to a possible future endpoint for the networks.

Looking forward to your feedback. I'd also like to wish everyone a Merry Christmas and only the very best for 2023.

Regards,

Jean-Luc

[^1]: *Nextbike* would call them *domains*.

### CODE REVIEW

File: pybikes/publibike.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        'company': ['PubliBike AG'],
+        'source': 'https://api.publibike.ch/v1/static/api.html'
+    }
+
+    def __init__(self, tag, meta, city_uid, hostname='api.publibike.ch',
+                 bbox=None):
+        super(Publibike, self).__init__(tag, meta)
+        self.url = BASE_URL.format(hostname=hostname)
+        self.uid = city_uid
+        self.bbox = bbox
```
Comment (eskerda) [inline pybikes/publibike.py:36]: Unused bbox argument

File: pybikes/publibike.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        stations = stations['stations']
+
+        # currently (Dezember 2022) there is no endpoint available to query only stations for 'city_uid'
+        # so we need to filter the data
+        stations = filter(lambda s: s['network']['id'] == self.uid, stations)
+
+        if self.bbox:
+            def getter(station):
+                lat, lng = station['latitude'], station['longitude']
+                return (float(lat), float(lng))
```
Comment (eskerda) [inline pybikes/publibike.py:58]: Not used

File: pybikes/publibike.py (hunk 1 above)
Comment (eUgEntOptIc44) [reply pybikes/publibike.py]: @eskerda was leftover from the nextbike code that I used as a base. Now removed.

File: pybikes/publibike.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2010-2022, eskerda <eskerda@gmail.com>
+# Copyright (C) 2022, eUgEntOptIc44 (https://github.com/eUgEntOptIc44)
+# Distributed under the AGPL license, see LICENSE.txt
+
+import re
+import json
+
+from .base import BikeShareSystem, BikeShareStation
+from pybikes.utils import PyBikesScraper, filter_bounds
```
Comment (eskerda) [inline pybikes/publibike.py:10]: remove unused import

File: pybikes/publibike.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+from pybikes.contrib import TSTCache
+
+__all__ = ['Publibike', 'PublibikeStation']
```
Comment (eskerda) [inline pybikes/publibike.py:13]: not needed

File: pybikes/publibike.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+        if 'city' in station:
+            self.extra['city'] = station['city']
+
+        self.bikes = 0
+        if "vehicles" in station:
+            self.bikes = len(station['vehicles'])
+
+        self.free = None
+        if "capacity" in station:
```
Comment (eskerda) [inline pybikes/publibike.py:81]: Does the contrary ever happen, as in, is `capacity` ever not in station?

File: pybikes/publibike.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            try:
+                self.extra['slots'] = int(station['capacity'])
+            except TypeError:
+                self.extra['slots'] = 0
```
Comment (eskerda) [inline pybikes/publibike.py:85]: Is there any instance where the `capacity` set on the station is not an integer? If that is not the case, let's not set up guards against conditions that do not happen, and let them fail if they do happen

File: pybikes/publibike.py [hunk 7]
```diff
@@ -78,11 +78,10 @@ def __init__(self, station):
...
-            try:
-                self.extra['slots'] = int(station['capacity'])
-            except TypeError:
-                self.extra['slots'] = 0
-
-            if self.extra['slots'] > 0:
-                self.free = self.extra['slots'] - self.bikes
+        try:
+            self.extra['slots'] = station['capacity']
+        except TypeError:
```
Comment (eskerda) [inline pybikes/publibike.py:83]: Why is this a type error?

```
self.extra['slots'] = station.get('capacity', 0)

File: pybikes/publibike.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+# Copyright (C) 2022-2023, eUgEntOptIc44 (https://github.com/eUgEntOptIc44)
+# Distributed under the AGPL license, see LICENSE.txt
+
+import json
+
+from .base import BikeShareSystem, BikeShareStation
+from pybikes.utils import PyBikesScraper
+from pybikes.contrib import TSTCache
+
+__all__ = ['Publibike', 'PublibikeStation']
```
Comment (eskerda) [inline pybikes/publibike.py:12]: Remove this line

File: pybikes/publibike.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+    meta = {
+        'system': 'PubliBike',
+        'company': ['PubliBike AG'],
+        'source': 'https://api.publibike.ch/v1/static/api.html'
+    }
+
+    def __init__(self, tag, meta, city_uid, hostname='api.publibike.ch'):
+        super(Publibike, self).__init__(tag, meta)
+        self.url = BASE_URL.format(hostname=hostname)
```
Comment (eskerda) [inline pybikes/publibike.py:32]: url is a constant and is always the same. Remove the `hostname` argument, remove `self.url` too, and change `BASE_URL` into

```
FEED_URL = 'https://api.publibike.ch/v1/public/partner/stations'
```

File: pybikes/publibike.py (hunk 8 above)
Comment (eUgEntOptIc44) [reply pybikes/publibike.py:12]: ok

File: pybikes/publibike.py (hunk 9 above)
Comment (eUgEntOptIc44) [reply pybikes/publibike.py:32]: Ok done

Comment (eUgEntOptIc44) [reply pybikes/publibike.py:32]: done

**Comment by eUgEntOptIc44** [comment]
Hi @eskerda,
Thanks for your patience. I tried to apply everyting. I'd be very grateful if you provide feedback on the current status of this PR. Is there anything left to change before merging?

File: pybikes/publibike.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+        self.bikes = 0
+        if "vehicles" in station:
+            self.bikes = len(station['vehicles'])
+
+        self.free = None
+        self.extra['slots'] = station.get('capacity', 0)
+
+        if self.extra['slots'] > 0:
+            self.free = self.extra['slots'] - self.bikes
```
Comment (eskerda) [inline pybikes/publibike.py:81]: No need to add these conditionals if the information is always there!

```
        self.name = station['name']
        self.latitude = float(station['latitude'])
        self.longitude = float(station['longitude'])
        self.extra = {
            'uid': station['id'],
            'address': station['address'],
            'zip': station['zip'],
            'city': station['city'],
            'slots': station['capacity'],
        }
        self.bikes = len(station['vehicles'])
        self.free = self.extra['slots'] - self.bikes
```

File: pybikes/publibike.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2010-2022, eskerda <eskerda@gmail.com>
+# Copyright (C) 2022-2023, eUgEntOptIc44 (https://github.com/eUgEntOptIc44)
+# Distributed under the AGPL license, see LICENSE.txt
+
+import json
+
+from .base import BikeShareSystem, BikeShareStation
+from pybikes.utils import PyBikesScraper
```
Comment (eskerda) [inline pybikes/publibike.py:9]:
```suggestion
from pybikes import BikeShareSystem, BikeShareStation, PyBikesScraper
```

**Review by eskerda — CHANGES_REQUESTED** [review]

File: pybikes/publibike.py (hunk 11 above)
Comment (eUgEntOptIc44) [reply pybikes/publibike.py:9]: ok done

File: pybikes/publibike.py (hunk 10 above)
Comment (eUgEntOptIc44) [reply pybikes/publibike.py:81]: applied

**Comment by eUgEntOptIc44** [comment]
Hi @eskerda I applied the requested changes.

**Comment by eskerda** [comment]
Good job @eUgEntOptIc44!

**Review by eskerda — APPROVED** [review]

**Pull request merged**